src/librustc/ty/structural_impls.rs

```diff
@@ -828,15 +825,25 @@ impl<'tcx> TypeFoldable<'tcx> for ty::ProjectionPredicate<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for ty::ExistentialProjection<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
-        ty::ExistentialProjection {
-            trait_ref: self.trait_ref.fold_with(folder),
-            item_name: self.item_name,
-            ty: self.ty.fold_with(folder),
-        }
+        // FIXME(tschottdorf): simplify.
+        let folded_ty = self.ty.fold_with(folder);
+        let trait_ref = self.trait_ref(folder.tcx());
+        let item_name = self.item_name(folder.tcx());
+        let folded_trait_ref = trait_ref.fold_with(folder);
+        ty::ExistentialProjection::from_ty_ref_and_name(
+            folder.tcx(),
+            folded_ty,
+            folded_trait_ref,
+            item_name,
+        )
     }
 
     fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.trait_ref.visit_with(visitor) || self.ty.visit_with(visitor)
+        // FIXME(tschottdorf): find/plumb a `tcx` here:
+        // self.trait_ref(tcx()).visit_with(visitor) || ...
+        ty::tls::with(|tcx|
+            self.trait_ref(tcx).visit_with(visitor) || self.ty.visit_with(visitor)
```

Comment:
`ExistentialTraitRef` doesn't go to `visit_trait_ref` so you don't have to do anything other than visit the substs.